File: artifacts/ht/refactor/dump.c
#include <ht.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct ht ht;
ht *ht_create(void);
void ht_destroy(ht *table);
void *ht_get(ht *table, const char *key);
const char *ht_set(ht *table, const char *key, void *value);
size_t ht_length(ht *table);
typedef struct 
{
  const char *key;
  void *value;
  ht *_table;
  size_t _index;
} hti;
hti ht_iterator(ht *table);
bool ht_next(hti *it);
typedef struct 
{
  char *key;
  void *value;
} ht_entry;
struct ht
{
  ht_entry *entries;
  size_t capacity;
  size_t length;
};
typedef struct 
{
  char *key;
  int value;
} item;
void exit_nomem(void)
{
  fprintf(stderr, "out of memory\n");
  exit(1);
}

int main(void)
{
  item items[] = {{"foo", 10}, {"bar", 42}, {"bazz", 36}, {"buzz", 7}, {"bob", 11}, {"jane", 100}, {"x", 200}};
  size_t num_items = (sizeof(items)) / (sizeof(item));
  ht *table = ht_create();
  if (table == 0)
  {
    exit_nomem();
  }
  for (int i = 0; i < num_items; i += 1)
  {
    if (ht_set(table, items[i].key, &items[i].value) == 0)
    {
      exit_nomem();
    }
  }

  for (int i = 0; i < table->capacity; i += 1)
  {
    if (table->entries[i].key != 0)
    {
      printf("index %d: key %s, value %d\n", i, table->entries[i].key, *((int *) table->entries[i].value));
    }
    else
    {
      printf("index %d: empty\n", i);
    }
  }

  return 0;
}

File: artifacts/ht/translate_ablation_1/src/dump.rs
use std::any::Any;
use std::boxed::Box;
use std::env;
use std::ffi::CStr;
use std::io::BufRead;
use std::io;
use translate_new::*;
pub fn main() {
    // Build some sample items (keys are Option<String> per translated types)
    let items = [
        Item { key: Some("foo".to_string()),  value: 10  },
        Item { key: Some("bar".to_string()),  value: 42  },
        Item { key: Some("bazz".to_string()), value: 36  },
        Item { key: Some("buzz".to_string()), value: 7   },
        Item { key: Some("bob".to_string()),  value: 11  },
        Item { key: Some("jane".to_string()), value: 100 },
        Item { key: Some("x".to_string()),    value: 200 },
    ];
    let num_items = items.len();

    // Create table (translated API returns Option<Ht>)
    let mut table = ht_create();
    if table.is_none() {
        exit_nomem();
    }

    // Insert: ht_set expects (&mut Ht, Option<&str>, Option<Box<dyn Any>>)
    for i in 0..num_items {
        let key_opt: Option<&str> = items[i].key.as_deref();
        let val_box = Some(Box::new(items[i].value) as Box<dyn std::any::Any>);
        if ht_set(table.as_mut().unwrap(), key_opt, val_box).is_none() {
            exit_nomem();
        }
    }

    // Dump table contents
    let table_ref = table.as_ref().unwrap();
    if let Some(ref entries) = table_ref.entries {
        for (i, entry) in entries.iter().enumerate() {
            if let Some(ref key_str) = entry.key {
                // value is Option<Box<dyn Any>>; downcast to i32 for display
                let shown_val = entry
                    .value
                    .as_ref()
                    .and_then(|b| b.as_ref().downcast_ref::<i32>())
                    .copied();
                match shown_val {
                    Some(v) => println!("index {}: key {}, value {}", i, key_str, v),
                    None => println!("index {}: key {}, value <non-i32>", i, key_str),
                }
            } else {
                println!("index {}: empty", i);
            }
        }
    } else {
        println!("hash table has no buckets (entries = None)");
    }
}
